Fix iOS Simulated Keyboard Events Crash (event-timing)

francovs opened the pull request:
#### 2f84d00fc31557a3e0a0bd1fc5dbe32a94a40e04
<pre>
Fix iOS Simulated Keyboard Events Crash (event-timing)
<a href="https://bugs.webkit.org/show_bug.cgi?id=300055">https://bugs.webkit.org/show_bug.cgi?id=300055</a>
<a href="https://rdar.apple.com/161854148">rdar://161854148</a>

Reviewed by Abrar Rahman Protyasha.

Avoids unsafe downcast of key{up/down/press} events to KeyboardEvent,
which was crashing on iOS simulated keyboard events. Crashes were
occurring in these two API tests:

- KeyboardInputTests.InsertDictationAlternativesSimulatingKeyboardInput
- KeyboardInputTests.InsertTextSimulatingKeyboardInput

(these tests are not currently being exercised by automation
because the feature is currently behind a feature flag)

* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::computeInteractionID):
(WebCore::LocalDOMWindow::finalizeEventTimingEntry):

Canonical link: <a href="https://commits.webkit.org/300968@main">https://commits.webkit.org/300968@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ff27360b2292b94e78c1b7923470bc323933e822

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124472 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44152 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34887 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131315 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76467 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126349 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44880 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52739 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94690 "Found 1 new test failure: fast/backgrounds/obscured-background-child-style-change.html (failure)") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76467 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127426 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35774 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111326 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75267 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/34710 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29487 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74794 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105526 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29713 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133980 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51355 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39180 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103166 "16 flakes 55 failures") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51760 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107548 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102953 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26215 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48320 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26578 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/48300 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51224 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57011 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50637 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53998 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52315 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2499,8 +2499,11 @@ EventTimingInteractionID LocalDOMWindow::computeInteractionID(Event& event, Even
-        ASSERT(event.isKeyboardEvent());
-        auto keyboardEvent = downcast<KeyboardEvent>(&event);
+        // Fails on iOS simulated keyboard events:
+        auto keyboardEvent = dynamicDowncast<KeyboardEvent>(&event);
```
```suggestion
        RefPtr keyboardEvent = dynamicDowncast<KeyboardEvent>(&event);
```

aprotyas commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2529,11 +2532,11 @@ EventTimingInteractionID LocalDOMWindow::computeInteractionID(Event& event, Even
-        if (!event.isInputEvent())
+        // Fails for events not related to text, such as checkbox toggling:
+        auto inputEvent = dynamicDowncast<InputEvent>(&event);
+        if (!inputEvent)
```
I think you also want to consult `!isInputEvent()`?

aprotyas replied:
right, @whsieh?

francovs replied:
These are the only definitions:

```C++
// Event.h :
virtual bool isInputEvent() const { return false; }

// InputEvent.h:
bool isInputEvent() const override { return true; }
```

Same for `isKeyboardEvent()`, `isMouseEvent()`, etc... since there's no extra logic we should probably prefer `is<TYPE>(event)` instead.

rreno replied:
`is<TYPE>` is implemented in terms of these member functions. (That's what the `SPECIALIZE_TYPE_TRAITS_EVENT` macro does)

aprotyas replied:
ah good catch

aprotyas commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2529,11 +2532,11 @@ EventTimingInteractionID LocalDOMWindow::computeInteractionID(Event& event, Even
-        // Return early for events not related to text, such as checkbox toggling:
-        if (!event.isInputEvent())
+        // Fails for events not related to text, such as checkbox toggling:
+        auto inputEvent = dynamicDowncast<InputEvent>(&event);
```
```suggestion
        RefPtr inputEvent = dynamicDowncast<InputEvent>(&event);
```

aprotyas commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2653,7 +2656,12 @@ void LocalDOMWindow::finalizeEventTimingEntry(PerformanceEventTimingCandidate& e
     case EventType::keydown: {
-        auto keyboardEvent = downcast<KeyboardEvent>(&event);
+        // Fails on iOS simulated keyboard events:
+        auto keyboardEvent = dynamicDowncast<KeyboardEvent>(&event);
```
```suggestion
        RefPtr keyboardEvent = dynamicDowncast<KeyboardEvent>(&event);
```

aprotyas commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2679,7 +2687,11 @@ void LocalDOMWindow::finalizeEventTimingEntry(PerformanceEventTimingCandidate& e
     }
     case EventType::keypress: {
-        auto keyboardEvent = downcast<KeyboardEvent>(&event);
+        auto keyboardEvent = dynamicDowncast<KeyboardEvent>(&event);
```
```suggestion
        RefPtr keyboardEvent = dynamicDowncast<KeyboardEvent>(&event);
```

aprotyas reviewed (approved):
```
- KeyboardInputTests.InsertDictationAlternativesSimulatingKeyboardInput
- KeyboardInputTests.InsertTextSimulatingKeyboardInput
```

For these two tests, do you mind writing variants with event timing enabled?

You can create a WKWebViewConfiguration with the feature enabled like this:

```
    RetainPtr configuration = [WKWebViewConfiguration _test_configurationWithTestPlugInClassName:@"WebProcessPlugInWithInternals" configureJSCForTesting:YES];

    for (_WKFeature *feature in [WKPreferences _features]) {
        if ([feature.key isEqualToString:@"EventTimingEnabled"])
            [[configuration preferences] _setEnabled:YES forFeature:feature];
    }
```

rreno commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2499,8 +2499,11 @@ EventTimingInteractionID LocalDOMWindow::computeInteractionID(Event& event, Even
     case EventType::keyup: {
-        ASSERT(event.isKeyboardEvent());
-        auto keyboardEvent = downcast<KeyboardEvent>(&event);
+        // Fails on iOS simulated keyboard events:
```
Why does this fail on iOS for simulated events? Ditto the other comments indicating failure elsewhere.

francovs replied:
Rewrote as follows: `// Simulated keyboard inputs such as dictation are not KeyboardEvent:`

francovs committed:
Fix iOS Simulated Keyboard Events Crash (event-timing)
https://bugs.webkit.org/show_bug.cgi?id=300055
rdar://161854148

Reviewed by Abrar Rahman Protyasha.

Avoids unsafe downcast of key{up/down/press} events to KeyboardEvent,
which was crashing on iOS simulated keyboard events. Crashes were
occurring in these two API tests:

- KeyboardInputTests.InsertDictationAlternativesSimulatingKeyboardInput
- KeyboardInputTests.InsertTextSimulatingKeyboardInput

(these tests are not currently being exercised by automation
because the feature is currently behind a feature flag)

* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::computeInteractionID):
(WebCore::LocalDOMWindow::finalizeEventTimingEntry):

Canonical link: https://commits.webkit.org/300968@main
